File: proj.android_studio/ee-x-core/src/main/java/com/ee/core/MessageBridge.java
package com.ee.core;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.ee.core.internal.JsonUtils;

import java.util.HashMap;
import java.util.Map;

import static com.google.common.truth.Truth.assertThat;

/**
 * Created by Zinge on 3/29/17.
 */

public class MessageBridge implements IMessageBridge {
    private static final Logger _logger = new Logger(MessageBridge.class.getName());

    /**
     * Registered handlers.
     */
    private final Map<String, MessageHandler> _handlers;

    private final Object _handlerLock;

    /**
     * Thread-safe singleton pattern.
     */
    private static class Holder {
        private static final MessageBridge Instance = new MessageBridge();
    }

    /**
     * Gets a message bridge instance.
     *
     * @return A message bridge instance.
     */
    public static IMessageBridge getInstance() {
        return Holder.Instance;
    }

    private MessageBridge() {
        _handlers = new HashMap<>();
        _handlerLock = new Object();
    }

    public boolean registerHandler(@NonNull String tag, @NonNull MessageHandler handler) {
        synchronized (_handlerLock) {
            if (_handlers.containsKey(tag)) {
                _logger.error("registerHandler: " + tag + " already exists!");
                return false;
            }
            _handlers.put(tag, handler);
            return true;
        }
    }

    public boolean registerAsyncHandler(@NonNull String tag, @NonNull AsyncMessageHandler handler) {
        return registerHandler(tag, message -> {
            Map<String, Object> dict = JsonUtils.convertStringToDictionary(message);
            assertThat(dict).isNotNull();

            String callbackTag = (String) dict.get("callback_tag");
            String originalMessage = (String) dict.get("message");
            assertThat(callbackTag).isNotNull();
            assertThat(originalMessage).isNotNull();

            handler.handle(originalMessage, callbackMessage ->
                callCpp(callbackTag, callbackMessage));
            return "";
        });
    }

    public boolean deregisterHandler(@NonNull String tag) {
        synchronized (_handlerLock) {
            if (!_handlers.containsKey(tag)) {
                _logger.error("deregisterHandler: " + tag + " doesn't exist!");
                return false;
            }
            _handlers.remove(tag);
            return true;
        }
    }

    @Nullable
    private MessageHandler findHandler(@NonNull String tag) {
        synchronized (_handlerLock) {
            if (!_handlers.containsKey(tag)) {
                return null;
            }
            return _handlers.get(tag);
        }
    }

    @NonNull
    public String callCpp(@NonNull String tag) {
        return callCpp(tag, "");
    }

    @NonNull
    public String callCpp(@NonNull String tag, @NonNull String message) {
        return callCppInternal(tag, message);
    }

    /**
     * Calls a handler from C++ with a message.
     *
     * @param tag     The unique tag of the handler.
     * @param message The message.
     * @return Reply message from C++.
     */
    @NonNull
    private native String callCppInternal(@NonNull String tag, @NonNull String message);

    /**
     * Calls a handler from Java with a message.
     *
     * @param tag     The unique tag of the handler.
     * @param message The message.
     * @return Reply message from Java.
     */
    @NonNull
    private String call(@NonNull String tag, @NonNull String message) {
        MessageHandler handler = findHandler(tag);
        if (handler == null) {
            _logger.error("call: " + tag + " doesn't exist!");
            return "";
        }
        return handler.handle(message);
    }

    /**
     * Statically calls a handler from Java with a message.
     *
     * @param tag     The unique tag of the handler.
     * @param message The message.
     * @return Reply message from Java.
     */
    @NonNull
    private static String staticCall(@NonNull String tag, @NonNull String message) {
        return Holder.Instance.call(tag, message);
    }
}

File: proj.android_studio/ee-x-core/src/main/java/com/ee/internal/MessageBridge.kt
package com.ee.internal

import com.ee.AsyncMessageHandler
import com.ee.ILogger
import com.ee.IMessageBridge
import com.ee.MessageHandler
import com.ee.PluginManager
import com.google.common.truth.Truth.assertThat
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import kotlinx.serialization.InternalSerializationApi
import kotlinx.serialization.Serializable

typealias MessageBridgeHandler = (tag: String, message: String) -> String

/**
 * Created by Zinge on 3/29/17.
 */
class MessageBridge(
    private val _logger: ILogger,
    private val _handler: MessageBridgeHandler)
    : IMessageBridge {
    companion object {
        private val kTag = MessageBridge::class.java.name
    }

    /**
     * Registered handlers.
     */
    private val _handlers: MutableMap<String, MessageHandler>
    private val _handlerLock: Any
    private val _scope = MainScope()

    init {
        _handlers = HashMap()
        _handlerLock = Any()
    }

    override fun registerHandler(tag: String, handler: MessageHandler): Boolean {
        synchronized(_handlerLock) {
            if (_handlers.containsKey(tag)) {
                _logger.error("$kTag: registerHandler: $tag already exists!")
                assertThat(false).isTrue()
                return false
            }
            _handlers[tag] = handler
            return true
        }
    }

    @InternalSerializationApi
    override fun registerAsyncHandler(tag: String, handler: AsyncMessageHandler): Boolean {
        return registerHandler(tag) { message ->
            @Serializable
            class Request(
                val callback_tag: String,
                val message: String
            )

            val request = deserialize<Request>(message)
            _scope.launch {
                val response = handler(request.message)
                callCpp(request.callback_tag, response)
            }
            ""
        }
    }

    override fun deregisterHandler(tag: String): Boolean {
        synchronized(_handlerLock) {
            if (!_handlers.containsKey(tag)) {
                _logger.error("$kTag: deregisterHandler: $tag doesn't exist!")
                assertThat(false).isTrue()
                return false
            }
            _handlers.remove(tag)
            return true
        }
    }

    private fun findHandler(tag: String): MessageHandler? {
        synchronized(_handlerLock) {
            return _handlers[tag]
        }
    }

    override fun call(tag: String, message: String): String {
        val handler = findHandler(tag)
        if (handler == null) {
            _logger.error("$kTag: call: $tag doesn't exist!")
            assertThat(false).isTrue()
            return ""
        }
        return handler(message)
    }

    override fun callCpp(tag: String): String {
        return callCpp(tag, "")
    }

    override fun callCpp(tag: String, message: String): String {
        return _handler(tag, message)
    }
}

internal external fun ee_callCppInternal(tag: String, message: String): String

@NativeThread
@Suppress("unused")
private fun ee_staticCall(tag: String, message: String): String {
    val bridge = PluginManager.getInstance().getBridge()
        ?: throw IllegalStateException("Bridge is null")
    return bridge.call(tag, message)
}